ts/config.ts

```diff
@@ -1,33 +1,65 @@
-import * as yaml from "js-yaml";
+import { fold } from "fp-ts/Either";
+import { pipe } from "fp-ts/function";
+import * as util from "util";
 import * as fs from "fs";
+import * as io_ts from "io-ts";
+import * as yaml from "js-yaml";
 import * as path from "path";
 import { logger } from "./logger";
 
-export interface web_greeter_config {
-  branding: {
-    background_images_dir: string;
-    logo_image: string;
-    user_image: string;
-  };
-  greeter: {
-    debug_mode: boolean;
-    detect_theme_errors: boolean;
-    screensaver_timeout: number;
-    secure_mode: boolean;
-    theme: string;
-    icon_theme: string | undefined;
-    time_language: string | undefined;
-  };
-  layouts: string[];
-  features: {
-    battery: boolean;
-    backlight: {
-      enabled: boolean;
-      value: number;
-      steps: number;
-    };
-  };
-}
+export const WEB_GREETER_CONFIG = io_ts.type({
+  branding: io_ts.type({
+    background_images_dir: io_ts.string,
+    logo_image: io_ts.string,
+    user_image: io_ts.string,
+  }),
+  greeter: io_ts.type({
+    debug_mode: io_ts.boolean,
+    detect_theme_errors: io_ts.boolean,
+    screensaver_timeout: io_ts.number,
+    secure_mode: io_ts.boolean,
+    theme: io_ts.string,
+    icon_theme: io_ts.union([io_ts.string, io_ts.null]),
+    time_language: io_ts.union([io_ts.string, io_ts.null]),
```

Comment:
Any reason why you didn't do this instead?

```suggestion
    icon_theme: io_ts.union([io_ts.string, io_ts.undefined]),
    time_language: io_ts.union([io_ts.string, io_ts.undefined]),
```

If you have null, when these values don't exist it will probably fail the `decode()`

Reply:
Have confirmed that this is the case, so `io_ts.undefined` is probably better here

Reply:
When `yaml` does the parse, it returns `null` instead of `undefined` on empty properties. So, it's better to handle null instead of undefined.

Reply:
Ah, in that case we should / could allow for both `null` and `undefined`, as if the line is missing from the config it will be undefined:

```suggestion
    icon_theme: io_ts.union([io_ts.string, io_ts.null, io_ts.undefined]),
    time_language: io_ts.union([io_ts.string, io_ts.null, io_ts.undefined]),
```